### ISSUE TITLE
Hanging workers after update to Redis 3.3.0

### ISSUE THREAD

**Issue by Mingan** [resque/resque#1452]
Yesterday, I updated redis-rb gem from 3.2.2 to 3.3.0 and started seeing workers to hang indefinitely (as processing a job). 
Our setup is MRI 2.1.5, resque 1.25.2, redis-namespace 1.5.2, resque-pool 0.6.0, Redis 2.2.12.

I've tried to narrow down the problem:
- Changing redis-rb gem version to 3.2.2 fixes it (= under full queue a worker hasn't hanged in an hour)
- Running `rake resque:workers` with the same number of workers still causes worker to hang some time (= it doesn't look like resque-pool issue)
- Running `rake resque:work` or `COUNT=1 rake resque:workers` seems to work ok (= no hanging worker after an hour of constant load)
- Sometimes first job gets stuck, sometimes it takes minutes or even tens of minutes
- It happens on different queues with different worker classes
- Looking at the log, the whole `perfom` function is executed

Cleaned up log of a case when the very first job got stuck:
```
I, [2016-04-19T14:05:34.249934 #7161]  INFO -- : Starting worker dev:7161:queue_name
D, [2016-04-19T14:05:34.250332 #7161] DEBUG -- : Registered signals
I, [2016-04-19T14:05:34.324168 #7161]  INFO -- : Running before_first_fork hooks
D, [2016-04-19T14:05:36.509096 #7161] DEBUG -- : Checking queue_name
D, [2016-04-19T14:05:36.512096 #7161] DEBUG -- : Found job on queue_name
I, [2016-04-19T14:05:36.512268 #7161]  INFO -- : got: (Job{queue_name} | CLS | ['some', 'arguments'])
D, [2016-04-19T14:05:36.513668 #7161] DEBUG -- : resque-1.25.2: Processing queue_name since 1461067536 [CLS]
I, [2016-04-19T14:05:36.513869 #7161]  INFO -- : Running before_fork hooks with [(Job{queue_name} | CLS | ['some', 'arguments'])]
D, [2016-04-19T14:05:36.521803 #7161] DEBUG -- : resque-1.25.2: Forked 7994 at 1461067536
```

Application code doesn't touch Redis connection, it calls `Resque.enqueue` at most.

Am I doing something wrong, or is there really a bug?

**Comment by ponny** [comment]
Same issue here.  It seems to stall on some jobs as well as adding them (caught by rack-timeout).  Most jobs were fine. 

Just did 3.3.0 upgrade.  Downgraded to 3.2.2 and it's fine again.

**Comment by erickromathp** [comment]
We had the same issue at a project we have at work, thanks for posting this issue.  We investigated for a couple of days until we realized redis was the issue.  

After upgrading to redis 3.3 the forked processes that resque workers create would finish successfully, however the worker process would stall and not take any more work, as if the forked process would still be running, but it was not (PID of forked process did not exist anymore).  

Downgrading to 3.2.2 fixed it as well.

**Comment by splittingred** [comment]
We experienced the same thing as well; filed in https://github.com/nevans/resque-pool/issues/156

Downgrading redis-rb back to 3.2.2 fixed the issue for us.

**Comment by jschneiderhan** [comment]
I've been debugging this issue as well, and I've gone down to the level of adding some custom logging statements to ruby, recompiling, and reviewing the debug logs once a thread hangs. I'm using resque 1.26 and the resque-pool gem. Here's a sample of the debug logs (note some of the these are not in the ruby source - I added them):

```
2448925:thread.c:885: pid:24807 ppid:24732 0x7f0f07cf2740 - thread_join (thid: 0x007f0ef5c16700)
2448926:thread.c:892: pid:24807 ppid:24732 0x7f0f07cf2740 - thread_join: right before rb_ensure call (thid: 0x007f0ef5c16700)
2448929:thread.c:848: pid:24807 ppid:24732 0x7f0f07cf2740 - thread_join_sleep: starting (thid: 0x007f0ef5c16700)
2448930:thread.c:852: pid:24807 ppid:24732 0x7f0f07cf2740 - thread_join_sleep: about to call sleep_forever (thid: 0x007f0ef5c16700)
2448931:thread.c:1062: pid:24807 ppid:24732 0x7f0f07cf2740 - sleep_forever: start (thid: 0x007f0f07cf2740, status:0, has_gvl:1)
2448932:thread.c:1064: pid:24807 ppid:24732 0x7f0f07cf2740 - sleep_forever: about to run first RUBY_VM_CHECK_INTS_BLOCKING (thid: 0x007f0f07cf2740, status:2, has_gvl:1)
2448933:thread.c:262: pid:24807 ppid:24732 0x7f0f07cf2740 - vm_check_ints_blocking: about to call rb_threadptr_execute_interrupts
2448934:thread.c:1998: pid:24807 ppid:24732 0x7f0f07cf2740 - rb_threadptr_execute_interrupts: executing interrupts (thid: 0x007f0f07cf2740, status:2, has_gvl:1)
2448935:thread.c:2020: pid:24807 ppid:24732 0x7f0f07cf2740 - rb_threadptr_execute_interrupts: setting thread status to THREAD_RUNNABLE (thid: 0x007f0f07cf2741, status:2, has_gvl:1)
2448936:thread.c:2023: pid:24807 ppid:24732 0x7f0f07cf2740 - rb_threadptr_execute_interrupts: about to exec signal 17 (thid: 0x007f0f07cf2740, status:0, has_gvl:1)
2448938:thread.c:1998: pid:24807 ppid:24732 0x7f0f07cf2740 - rb_threadptr_execute_interrupts: executing interrupts (thid: 0x007f0f07cf2740, status:0, has_gvl:1)
2448939:thread.c:2057: pid:24807 ppid:24732 0x7f0f07cf2740 - rb_threadptr_execute_interrupts: timer interrupt (thid: 0x007f0f07cf2740, status:0, has_gvl:1)
2448940:thread_pthread.c:101: pid:24732 ppid:17197 0x7f0f07cf2740 - gvl_acquiredtthread.c:1278: pid:24732 ppid:17197 0x7f0f07cf2740 - leave blocking region (0x007f0fthread_pthread.c:117: pid:24807 ppid:24732 0x7f0f07cf2740 - gvl_released
2448959:thread.c:650: pid:24807 ppid:24732 0x7f0ef5c16700 - thread_join: about to call rb_threadptr_interrupt (thid: 0x007f0f07cf2740, status:0, has_gvl:1)
2448962:thread.c:653: pid:24807 ppid:24732 0x7f0ef5c16700 - thread_join: processing join_list (thid: 0x007f0f07cf2740, status:0, has_gvl:1)
2448964:thread_pthread.c:101: pid:24807 ppid:24732 0x7f0f07cf2740 - gvl_acquired
2448965:thread.c:1239: pid:24807 ppid:24732 0x7f0f07cf2740 - rb_thread_schedule/switch done
2448966:thread.c:2025: pid:24807 ppid:24732 0x7f0f07cf2740 - rb_threadptr_execute_interrupts: done exec'ing signal 17 (thid: 0x007f0f07cf2740, status:0, has_gvl:1)
2448967:thread.c:2028: pid:24807 ppid:24732 0x7f0f07cf2740 - rb_threadptr_execute_interrupts: setting thread status BACK to prev_status (thid: 0x007f0f07cf2740, status:2, has_gvl:1)
2448968:thread.c:2057: pid:24807 ppid:24732 0x7f0f07cf2740 - rb_threadptr_execute_interrupts: timer interrupt (thid: 0x007f0f07cf2740, status:2, has_gvl:1)
2448969:thread.c:1230: pid:24807 ppid:24732 0x7f0f07cf2740 - rb_thread_schedule

```

This is what I think is happening. Thread 0x7f0f07cf2740 is joining thread 0x7f0ef5c16700. Thread 0x7f0ef5c16700 completes it's work and sends a SIGCHLD to 0x7f0f07cf2740. 0x7f0f07cf2740 starts executing the signal handler (defined in [resque-pool](https://github.com/nevans/resque-pool/blob/master/lib/resque/pool.rb#L165)), during which it's thread status is temporarily set to THREAD_RUNNABLE. While it's in the middle of executing a nested timer interrupt occurs, which causes the gvl to be released and the ruby thread scheduler to be invoked. The thread scheduler  happens to pick 0x7f0ef5c16700 which at that point decides to [process it's join_list](https://github.com/ruby/ruby/blob/v2_3_1/thread.c#L649), but because thread 0x7f0f07cf2740's status has been temporarily set to THREAD_RUNNABLE, it is skipped. When the thread scheduler again picks 0x7f0f07cf2740 to execute, it finishes handling the signal, sets it's status bck to THREAD_STOPPED, and goes to sleep, waiting for 0x7f0ef5c16700 to process it's join_list and wake it up (not realizing it has already been skipped).

Anyway, I'm not 100% sure that's a correct assessment, but I've been looking at this for awhile and figured I'd share it somewhere :). I'm still not completely sure where the problem is, so I'll probably continue digging into it when I have time over the next few days. Looking at how ruby executes interrupts, I"m not sure how the timer interrupt can be executed while in the middle of running the SIGCHLD signal handler, unless something in the signal handler is calling a blocking operation, which on the surface, the resque-pool signal handler does not (it just calls [https://github.com/nevans/resque-pool/blob/master/lib/resque/pool.rb#L170](write_nonblock) on a pipe.).

I hope some of that made sense :/

**Comment by robertIngrum** [comment]
@jschneiderhan that definitely seems to make sense.  I'm about to start digging into this as well since we'd really like to be on the latest version of Redis, but unfortunately don't have a ton of background with resque's internals.  Have you had a chance to make any new progress so I'm not retracing steps?

**Comment by jschneiderhan** [comment]
@robertIngrum no, sorry. We ended up moving to sidekiq before I looked into the issue further.
